language-plutus-core/stdlib/Language/PlutusCore/StdLib/Meta/Data/Tuple.hs

```diff
@@ -26,15 +26,15 @@ import           Control.Lens.Indexed      (ifor, itraverse)
 import           Data.Traversable
 
 -- | A Plutus Core tuple.
-data Tuple ann = Tuple
+data Tuple term ann = Tuple
     { _tupleElementTypes :: [Type TyName ann]     -- ^ The types of elements of a tuple.
-    , _tupleTerm         :: Term TyName Name ann  -- ^ A term representation of the tuple.
+    , _tupleTerm         :: term ann              -- ^ A term representation of the tuple.
```

Comment:
We've lost something here - the type above uses `TyName` but we're not asserting that this term does.

Reply:
Yes - I'm conflicted about this. Ideally one would add the constraint `TermLike term TyName Name` to the type declaration, except that's sort of broken in Haskell. I fixed it with a GADT (I didn't know GADTs could feature record syntax).

Reply:
why not have the `term` parameter be `Term` rather than `Term TyNameName`?

Reply:
Not really an issue with the new version, but I preferred not to fix `tyname` and `name` to `TyName` and `Name` in the declaration of `TermLike` since presumably we want the generality - and I figured that we might still want them as parameters to `TermLike` since one could conceivably switch to one of those clever representations of terms that use De Bruijn indices and track bindings at the type level, which would mean the corresponding `Term` type is no longer polymorphic on the type of names.

Reply:
> We've lost something here - the type above uses `TyName` but we're not asserting that this term does.

I'd say it's not too important and the GADT is an overkill. But YMMV.

> Not really an issue with the new version, but I preferred not to fix `tyname` and `name` to `TyName` and `Name` in the declaration of `TermLike` since presumably we want the generality - and I figured that we might still want them as parameters to `TermLike` since one could conceivably switch to one of those clever representations of terms that use De Bruijn indices and track bindings at the type level, which would mean the corresponding `Term` type is no longer polymorphic on the type of names.

I agree with this reasoning.

Reply:
I just meant in this context the `term` parameter could be `* -> * -> *` or something. I agree in the general case.

Reply:
Current solution is fine.